distributed/scheduler.py

```diff
@@ -4963,3 +4964,52 @@ def __init__(self, task, last_worker):
         super(KilledWorker, self).__init__(task, last_worker)
         self.task = task
         self.last_worker = last_worker
+
+
+class WorkerStatus(SchedulerPlugin):
+    """
+    An extension and plugin to share worker status with a remote observer
+
+    This is used in cluster managers to keep updated about the status of the
+    scheduler.
+    """
+
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        self.scheduler.add_plugin(self)
+        self.scheduler.extensions["cluster-status"] = self
+
+        self.scheduler.handlers.update(
+            {"subscribe-worker-status": self.subscribe_worker_status}
+        )
+
+        self.bcomm = BatchedSend(interval="5ms", loop=scheduler.loop)
+
+    async def subscribe_worker_status(self, comm=None):
+        self.bcomm.start(comm)
+        ident = self.scheduler.identity()
+        for v in ident["workers"].values():
+            del v["metrics"]
+            del v["last_seen"]
+        return ident
+
+    def add_worker(self, worker=None, **kwargs):
+        if self.started:
+            ident = self.scheduler.workers[worker].identity()
+            del ident["metrics"]
+            del ident["last_seen"]
+            self.bcomm.send(["add", {"workers": {worker: ident}}])
```

Comment:
A few questions (not familiar with `BatchedSend` objects, and skimming their implementation didn't clarify things):

- If no client has subscribed to this stream, what happens here with this `send` call?
- What happens if two different clients call `subscribe-worker-status`? I can't tell for sure, but it looks like this would likely cause problems, since it would replace the existing `comm`?
- What happens if the client disconnects after subscribing?

If this plugin is all no-ops when no client has subscribed, and works with multiple subscribers, then I think this is fine to always have in `DEFAULT_EXTENSIONS`, otherwise it may be better to have off.

Reply:
> If no client has subscribed to this stream, what happens here with this send call?

The fact that we're in this method means that someone is subscribing.

> What happens if two different clients call subscribe-worker-status?

Good point.  After a quick fix (I'll push a bit later this afternoon), there will be one plugin per connection

It's also worth noting that, currently, this is only being used by Cluster objects, so I would only expect one connection per scheduler.

> What happens if the client disconnects after subscribing?

The send calls would generate errors, which would start to fill logs.

It's worth noting again though that clients aren't likely to call this, only cluster managers (or at least that's the intent).

I can resolve this though.

Maybe a general question, in the architecture of Dask-Yarn, is it likely that two cluster managers might both connect up to the same scheduler? (I remember something like this being possible).

Reply:
I think that, more broadly, my question is:

*"Would it make sense for YarnCluster to also provide a `scheduler_info` attribute?"*

I'm building functionality on top of this and want to ensure that that functionality can be useful generally both for SpecCluster and YarnCluster, assuming that those two implementations remain separate.

Reply:
What functionality are you building on top of this? This could be done for YarnCluster, sure. Multiple cluster manager could be attached to the same cluster, but that's not the normal mode of operation.

The questions I'm asking above are just general "is this plugin robust" questions.

Reply:
> What functionality are you building on top of this?

-  I'd like to make the widget depend on it, rather than assuming that a Scheduler object was present
-  Dask-Labextension: https://github.com/dask/dask-labextension/pull/73
-  I'll also probably want to use it in Adaptive so that we have a sense of 
    1. what the user asked for (`SpecCluster.worker_spec`)
    2. what we've successfully requested (`SpecCluster.workers`)
    3.  what has actually arrived (`Cluster.scheduler_info["workers"]`)

    Though I suspec that we could also get by with just 2 and 3, which are probably more generic.

Reply:
> The questions I'm asking above are just general "is this plugin robust" questions.

Yup.  And that's appreciated.